clang/lib/DPCT/UserDefinedRules/UserDefinedRules.cpp

```diff
@@ -344,17 +344,141 @@ MetaRuleObject::PatternRewriter::PatternRewriter(
+// directive with the contents of the referenced file.
+std::unique_ptr<llvm::MemoryBuffer>
+readYAMLFile(const llvm::StringRef &RuleFilePath) {
+  // Load the rule file into a MemoryBuffer
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> Buffer =
+      llvm::MemoryBuffer::getFile(RuleFilePath);
+  if (!Buffer) {
+    llvm::errs() << "Error: failed to read " << RuleFilePath << ": "
+                 << Buffer.getError().message() << "\n";
+    clang::dpct::ShowStatus(MigrationErrorInvalidRuleFilePath);
+    dpctExit(MigrationErrorInvalidRuleFilePath);
+  }
+
+  // Get the directory path of the rule file.
+  llvm::SmallString<128> DirectoryPath(RuleFilePath);
+  llvm::sys::path::remove_filename(DirectoryPath);
+
+  // Iterate over the input line by line.
+  std::stringstream Output, IncRuleFilePath;
+  const std::string IncDirective = "!include";
+
+  size_t Idx = 0;
+  bool IncDirectiveFound = false;
+  bool SkipWhiteSpaces = false;
+  bool IsLineBeginning = true;
+
+  llvm::StringRef BuffContent = std::move(*Buffer)->getBuffer();
+  const auto BuffSize = BuffContent.size();
+  while (Idx < BuffSize) {
+    unsigned char Ch = BuffContent[Idx];
+
+    // Skip white spaces at the beginning of the line if it contains a directive
+    if (IsLineBeginning) {
+      auto i = Idx;
+      auto c = Ch;
+
+      // lookahead for "!" directive after white spaces
+      for (; i < BuffContent.size(); i++) {
+        c = BuffContent[i];
+
+        // Stop at new line or first non-white space character
+        if (c == '\n' || !std::isspace(c)) {
+          break;
+        }
+      }
+
+      // Check if the line starts with a directive
+      if (c == '!') {
+        // Move Idx to the beginning of directive
+        Idx = i;
+
+        // Check if the directive is "!include"
+        if (!IncDirectiveFound && Idx + IncDirective.length() <= BuffSize &&
+            BuffContent.substr(Idx, IncDirective.length()) == IncDirective) {
+          // Move Idx to the end of directive
+          Idx += IncDirective.length();
+          IncDirectiveFound = true;
+          SkipWhiteSpaces = true;
+        }
+
+        // Update current character
+        Ch = BuffContent[Idx];
+      }
+
+      IsLineBeginning = false;
+    }
+
+    // Skip return carriage character
+    if (Ch == '\r') {
+      Idx++;
+      continue;
+    }
+
+    // Process IncRuleFilePath at end of the line
+    if (Ch == '\n') {
+      if (IncDirectiveFound) {
+        auto IncRuleFilePathStr = IncRuleFilePath.str();
+
+        if (!IncRuleFilePathStr.empty()) {
+          // Find the absolute path for the included rule file path
+          llvm::SmallString<128> IncRuleFileAbsPath = DirectoryPath;
+          llvm::sys::path::append(IncRuleFileAbsPath, IncRuleFilePathStr);
+
+          // Recursively process the included file
+          if (llvm::sys::fs::exists(IncRuleFileAbsPath)) {
+            Output << readYAMLFile(IncRuleFileAbsPath.str())->getBuffer().str();
```

Comment:
let's stop in case,  or else dpct will hang here.
fileA:  !include fileB 
fileB: !include  fileA

Reply:
Added a DenseSet to handle this scenario